File: target-blaster.js
function spawnTarget () {
    if (targetLife > 1000) {
        randX = randint(0, 4)
        randY = randint(0, 4)
    } else {
        if (gun.get(LedSpriteProperty.X) == 4) {
            randX = gun.get(LedSpriteProperty.X) + randint(-1, 0)
        } else if (gun.get(LedSpriteProperty.X) == 0) {
            randX = gun.get(LedSpriteProperty.X) + randint(0, 1)
        } else {
            randX = gun.get(LedSpriteProperty.X) + randint(-1, 1)
        }
        if (gun.get(LedSpriteProperty.Y) == 4) {
            randY = gun.get(LedSpriteProperty.Y) + randint(-1, 0)
        } else if (gun.get(LedSpriteProperty.X) == 0) {
            randY = gun.get(LedSpriteProperty.Y) + randint(0, 1)
        } else {
            randY = gun.get(LedSpriteProperty.Y) + randint(-1, 1)
        }
    }
    target = [
    game.createSprite(randX, randY - 1),
    game.createSprite(randX - 1, randY),
    game.createSprite(randX + 1, randY),
    game.createSprite(randX, randY + 1)
    ]
    if (randY - 1 == -1) {
        target[0].delete()
    }
    if (randX - 1 == -1) {
        target[1].delete()
    }
    if (randX + 1 == 5) {
        target[2].delete()
    }
    if (randY + 1 == 5) {
        target[3].delete()
    }
    elapsedTarget = 0
    elapsed = input.runningTime()
}
input.onButtonPressed(Button.A, function () {
    startPushed = true
    if (!(gameStart)) {
        for (let index = 0; index <= 2; index++) {
            basic.clearScreen()
            music.playTone(880, music.beat(BeatFraction.Quarter))
            basic.showNumber(3 - index)
            basic.pause(280)
        }
        gameStart = true
        combo = 0
        gun = game.createSprite(0, 0)
        targetLife = 5000
        spawnTarget()
    }
})
function judgePlayer (bool: boolean) {
    if (bool) {
        music.playSoundEffect(music.createSoundEffect(WaveShape.Square, 1, 1980, 255, 0, 300, SoundExpressionEffect.None, InterpolationCurve.Curve), SoundExpressionPlayMode.InBackground)
        game.addScore(1)
        combo += 1
        if (combo != 0 && combo % 5 == 0) {
            music.playSoundEffect(music.createSoundEffect(WaveShape.Square, 2559, 3674, 255, 0, 300, SoundExpressionEffect.None, InterpolationCurve.Logarithmic), SoundExpressionPlayMode.InBackground)
            game.addLife(1)
        }
        if (game.score() != 0 && game.score() % 10 == 0) {
            music.playTone(587, music.beat(BeatFraction.Eighth))
            music.playTone(622, music.beat(BeatFraction.Eighth))
            music.playTone(659, music.beat(BeatFraction.Eighth))
            if (targetLife <= 1000) {
                targetLife += -100
            } else if (targetLife <= 100) {
                targetLife += targetLife / 2
            } else {
                targetLife += -1000
            }
        }
    } else {
        music.playSoundEffect(music.createSoundEffect(WaveShape.Square, 286, 1, 255, 0, 300, SoundExpressionEffect.None, InterpolationCurve.Curve), SoundExpressionPlayMode.InBackground)
        game.removeLife(1)
        combo = 0
    }
    for (let 값 of target) {
        값.delete()
    }
    spawnTarget()
}
// reset for emulator
input.onButtonPressed(Button.AB, function () {
    control.reset()
})
let combo = 0
let elapsed = 0
let elapsedTarget = 0
let target: game.LedSprite[] = []
let gun: game.LedSprite = null
let randY = 0
let randX = 0
let targetLife = 0
let startPushed = false
let gameStart = false
gameStart = false
startPushed = false
let joystickReset = [pins.analogReadPin(AnalogPin.P1), pins.analogReadPin(AnalogPin.P0)]
game.setLife(3)
basic.showString("Target Blaster!!")
basic.forever(function () {
    if (game.isGameOver()) {
        if (gameStart) {
            gameStart = false
            music.playSoundEffect(music.createSoundEffect(WaveShape.Square, 2559, 1, 255, 0, 1000, SoundExpressionEffect.Tremolo, InterpolationCurve.Logarithmic), SoundExpressionPlayMode.UntilDone)
        }
    }
    if (gameStart) {
        gun.set(LedSpriteProperty.Brightness, 255)
        basic.pause(500)
        gun.set(LedSpriteProperty.Brightness, 50)
        basic.pause(500)
    }
})
basic.forever(function () {
    if (gameStart) {
        elapsedTarget = input.runningTime() - elapsed
        // shoot
        if (pins.digitalReadPin(DigitalPin.P2) == 1 || input.buttonIsPressed(Button.A) || input.buttonIsPressed(Button.B)) {
            music.playSoundEffect(music.createSoundEffect(WaveShape.Square, 1600, 1, 255, 0, 300, SoundExpressionEffect.None, InterpolationCurve.Curve), SoundExpressionPlayMode.InBackground)
            if (gun.get(LedSpriteProperty.X) == randX && gun.get(LedSpriteProperty.Y) == randY) {
                judgePlayer(true)
            } else {
                judgePlayer(false)
            }
        }
        if (elapsedTarget >= targetLife) {
            judgePlayer(false)
        }
    }
})
basic.forever(function () {
    if (gameStart) {
        if (pins.analogReadPin(AnalogPin.P1) > joystickReset[0] + 100) {
            gun.change(LedSpriteProperty.X, 1)
        } else if (pins.analogReadPin(AnalogPin.P1) < joystickReset[0] - 100) {
            gun.change(LedSpriteProperty.X, -1)
        }
        if (pins.analogReadPin(AnalogPin.P0) > joystickReset[1] + 100) {
            gun.change(LedSpriteProperty.Y, 1)
        } else if (pins.analogReadPin(AnalogPin.P0) < joystickReset[1] - 100) {
            gun.change(LedSpriteProperty.Y, -1)
        }
        basic.pause(200)
    }
})

File: target-blaster.py
def spawnTarget():
    global randX, randY, target, elapsedTarget, elapsed
    if targetLife > 1000:
        randX = randint(0, 4)
        randY = randint(0, 4)
    else:
        if gun.get(LedSpriteProperty.X) == 4:
            randX = gun.get(LedSpriteProperty.X) + randint(-1, 0)
        elif gun.get(LedSpriteProperty.X) == 0:
            randX = gun.get(LedSpriteProperty.X) + randint(0, 1)
        else:
            randX = gun.get(LedSpriteProperty.X) + randint(-1, 1)
        if gun.get(LedSpriteProperty.Y) == 4:
            randY = gun.get(LedSpriteProperty.Y) + randint(-1, 0)
        elif gun.get(LedSpriteProperty.X) == 0:
            randY = gun.get(LedSpriteProperty.Y) + randint(0, 1)
        else:
            randY = gun.get(LedSpriteProperty.Y) + randint(-1, 1)
    target = [game.create_sprite(randX, randY - 1),
        game.create_sprite(randX - 1, randY),
        game.create_sprite(randX + 1, randY),
        game.create_sprite(randX, randY + 1)]
    if randY - 1 == -1:
        target[0].delete()
    if randX - 1 == -1:
        target[1].delete()
    if randX + 1 == 5:
        target[2].delete()
    if randY + 1 == 5:
        target[3].delete()
    elapsedTarget = 0
    elapsed = input.running_time()

def on_button_pressed_a():
    global startPushed, gameStart, combo, gun, targetLife
    startPushed = True
    if not (gameStart):
        for index in range(3):
            basic.clear_screen()
            music.play_tone(880, music.beat(BeatFraction.QUARTER))
            basic.show_number(3 - index)
            basic.pause(280)
        gameStart = True
        combo = 0
        gun = game.create_sprite(0, 0)
        targetLife = 5000
        spawnTarget()
input.on_button_pressed(Button.A, on_button_pressed_a)

def judgePlayer(bool2: bool):
    global combo, targetLife
    if bool2:
        music.play_sound_effect(music.create_sound_effect(WaveShape.SQUARE,
                1,
                1980,
                255,
                0,
                300,
                SoundExpressionEffect.NONE,
                InterpolationCurve.CURVE),
            SoundExpressionPlayMode.IN_BACKGROUND)
        game.add_score(1)
        combo += 1
        if combo != 0 and combo % 5 == 0:
            music.play_sound_effect(music.create_sound_effect(WaveShape.SQUARE,
                    2559,
                    3674,
                    255,
                    0,
                    300,
                    SoundExpressionEffect.NONE,
                    InterpolationCurve.LOGARITHMIC),
                SoundExpressionPlayMode.IN_BACKGROUND)
            game.add_life(1)
        if game.score() != 0 and game.score() % 10 == 0:
            music.play_tone(587, music.beat(BeatFraction.EIGHTH))
            music.play_tone(622, music.beat(BeatFraction.EIGHTH))
            music.play_tone(659, music.beat(BeatFraction.EIGHTH))
            if targetLife <= 1000:
                targetLife += -100
            elif targetLife <= 100:
                targetLife += targetLife / 2
            else:
                targetLife += -1000
    else:
        music.play_sound_effect(music.create_sound_effect(WaveShape.SQUARE,
                286,
                1,
                255,
                0,
                300,
                SoundExpressionEffect.NONE,
                InterpolationCurve.CURVE),
            SoundExpressionPlayMode.IN_BACKGROUND)
        game.remove_life(1)
        combo = 0
    for 값 in target:
        값.delete()
    spawnTarget()
# reset for emulator

def on_button_pressed_ab():
    control.reset()
input.on_button_pressed(Button.AB, on_button_pressed_ab)

combo = 0
elapsed = 0
elapsedTarget = 0
target: List[game.LedSprite] = []
gun: game.LedSprite = None
randY = 0
randX = 0
targetLife = 0
startPushed = False
gameStart = False
gameStart = False
startPushed = False
joystickReset = [pins.analog_read_pin(AnalogPin.P1),
    pins.analog_read_pin(AnalogPin.P0)]
game.set_life(3)
basic.show_string("Target Blaster!!")

def on_forever():
    global gameStart
    if game.is_game_over():
        if gameStart:
            gameStart = False
            music.play_sound_effect(music.create_sound_effect(WaveShape.SQUARE,
                    2559,
                    1,
                    255,
                    0,
                    1000,
                    SoundExpressionEffect.TREMOLO,
                    InterpolationCurve.LOGARITHMIC),
                SoundExpressionPlayMode.UNTIL_DONE)
    if gameStart:
        gun.set(LedSpriteProperty.BRIGHTNESS, 255)
        basic.pause(500)
        gun.set(LedSpriteProperty.BRIGHTNESS, 50)
        basic.pause(500)
basic.forever(on_forever)

def on_forever2():
    global elapsedTarget
    if gameStart:
        elapsedTarget = input.running_time() - elapsed
        # shoot
        if pins.digital_read_pin(DigitalPin.P2) == 1 or input.button_is_pressed(Button.A) or input.button_is_pressed(Button.B):
            music.play_sound_effect(music.create_sound_effect(WaveShape.SQUARE,
                    1600,
                    1,
                    255,
                    0,
                    300,
                    SoundExpressionEffect.NONE,
                    InterpolationCurve.CURVE),
                SoundExpressionPlayMode.IN_BACKGROUND)
            if gun.get(LedSpriteProperty.X) == randX and gun.get(LedSpriteProperty.Y) == randY:
                judgePlayer(True)
            else:
                judgePlayer(False)
        if elapsedTarget >= targetLife:
            judgePlayer(False)
basic.forever(on_forever2)

def on_forever3():
    if gameStart:
        if pins.analog_read_pin(AnalogPin.P1) > joystickReset[0] + 100:
            gun.change(LedSpriteProperty.X, 1)
        elif pins.analog_read_pin(AnalogPin.P1) < joystickReset[0] - 100:
            gun.change(LedSpriteProperty.X, -1)
        if pins.analog_read_pin(AnalogPin.P0) > joystickReset[1] + 100:
            gun.change(LedSpriteProperty.Y, 1)
        elif pins.analog_read_pin(AnalogPin.P0) < joystickReset[1] - 100:
            gun.change(LedSpriteProperty.Y, -1)
        basic.pause(200)
basic.forever(on_forever3)
